sklearn/svm/src/liblinear/linear.cpp

```diff
@@ -1896,7 +1897,7 @@ static int solve_l1r_lr(
 		if(newton_iter == 0)
 			Gnorm1_init = Gnorm1_new;
 
-		if(Gnorm1_new <= eps*Gnorm1_init)
+		if(Gnorm1_new <= eps*Gnorm1_init || QP_no_change >= 10)
```

Comment:
where is the 10 coming from?

Reply:
Random value from Poisson(10).
It breaks the outer-loop if the inner-loop does not update anything 10 times in a row.